[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148736

Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

williamwen42 committed:
[dynamo] add ignore_intentional_skips (i.e. force_inline)

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add ignore_intentional_skips (i.e. force_inline)"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"


Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"


Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"


Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"


Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"


Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"


Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"


Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"


Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"


Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jansel commented on torch/_dynamo/external_utils.py:
```diff
@@ -187,3 +187,35 @@ def call_module_hooks_from_backward_state(
+
+
+# used by dont_skip_tracing decorator to ignore trace rules recursively
+_ignore_skip_function_variable = False
```
Can this be done without using global state?

williamwen42 replied:
How can we implement this without a global in the case of a skipped function? e.g.
```python
@dont_skip_tracing
def f():
    # do something that forces a skipped frame
    g()

def g():
    ...
```

We need to somehow remember during our call to `g` that `dont_skip_tracing` is still active.

My idea for doing this without a global would be to take advantage of wrapper detection (i.e. something similar to what I'm doing on the non-recursive disable PR). We can propagate `dont_skip_tracing` when handling normal graph breaks by wrapping the resume function, but I don't think this approach will work on the above case.

williamwen42 replied:
Do you think an approach of looking for a `dont_skip_tracing` wrapper function in the traceback would be better?

williamwen42 committed:
Update on "[dynamo] add recursive-only dont_skip_tracing (i.e. force_inline)"


Implementation details:
- `external_utils._ignore_skip_function_variable` is a global that toggles whether we should ignore most skip rules.
  - When active, `trace_rules.lookup_inner` will return `UserFunctionVariable` (i.e. trace normally) instead of `SkipFunctionVariable` (skip tracing), unless the function is located in `torch._dynamo` (since tracing into certain `torch._dynamo` modules, especially `torch._dynamo.eval_frame`, results in bad tracing behaviors). 
  - `SkipFunctionVariable.call_function` will also attempt to sourceless build its function and run `call_function` on that. This needs to be done since the `SkipFunctionVariable` may have been loaded before `_ignore_skip_function_variable` is active.
- `dont_skip_tracing(recursive=True)` wraps around a function and toggles `_ignore_skip_function_variable` around the function call.
- When Dynamo traces `_set_ignore_skip_function_variable`, it records the global side effect change to `_ignore_skip_function_variable` but also sets it for real in the current process.

Notes on correctness: We must ensure that our rules for setting `_ignore_skip_function_variable` are composable with graph breaks and eval_frame/caching behavior. In particular:
- `_ignore_skip_function_variable` should be set for real so that we do the correct thing in the case of non-inlined nested calls
- `_ignore_skip_function_variable` should be set when tracing so that we do the correct thing in the case of inlining

We follow these invariants in order to ensure correct behavior:
- `_ignore_skip_function_variable` should be set before and unset after the function call for real
- Dynamo-generated bytecode should produce the same results and have the same side effects as running the original code eagerly
- Dynamo can change global state in whatever way it wishes during tracing, as long as the global state is reset after tracing.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented:
Superceded by traceback approach
